lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1022 @@
+    for (auto &dep : fwd.dependencies) {
+      deps.push_back(
+        FileDependency::modTimeBased(
+          dep.path, dep.size, dep.lastModificationTime));
+    }
+    return dependenciesAreUpToDate(deps);
+  }
+
+  Optional<StringRef>
+  computePrebuiltModulePath(llvm::SmallString<256> &scratch) {
+    namespace path = llvm::sys::path;
+    StringRef sdkPath = ctx.SearchPathOpts.SDKPath;
+
+    // Check if the interface file comes from the SDK
+    if (sdkPath.empty() || !hasPrefix(path::begin(interfacePath),
+                                      path::end(interfacePath),
+                                      path::begin(sdkPath),
+                                      path::end(sdkPath)))
+      return None;
+
+    // Assemble the expected path: $PREBUILT_CACHE/Foo.swiftmodule or
+    // $PREBUILT_CACHE/Foo.swiftmodule/arch.swiftmodule. Note that there's no
+    // cache key here.
+    scratch.append(prebuiltCacheDir);
+
+    // FIXME: Would it be possible to only have architecture-specific names
+    // here? Then we could skip this check.
+    StringRef inParentDirName =
+      path::filename(path::parent_path(interfacePath));
+    if (path::extension(inParentDirName) == ".swiftmodule") {
+      assert(path::stem(inParentDirName) == moduleName);
+      path::append(scratch, inParentDirName);
+    }
+    path::append(scratch, path::filename(modulePath));
+
+    return scratch.str();
+  }
+
+  /// Finds the most appropriate .swiftmodule, whose dependencies are up to
+  /// date, that we can load for the provided .swiftinterface file.
+  llvm::ErrorOr<DiscoveredModule> discoverUpToDateModuleForInterface(
+    StringRef modulePath, StringRef cachedOutputPath,
+    SmallVectorImpl<FileDependency> &deps) {
+    auto notFoundError =
+      std::make_error_code(std::errc::no_such_file_or_directory);
+
+    // Keep track of whether we should attempt to load a .swiftmodule adjacent
+    // to the .swiftinterface.
+    bool shouldLoadAdjacentModule = true;
+
+    switch (loadMode) {
+    case ModuleLoadingMode::OnlyParseable:
+      // Always skip both the caches and adjacent modules, and always build the
+      // parseable interface.
+      return notFoundError;
+    case ModuleLoadingMode::PreferParseable:
+      // If we're in the load mode that prefers .swiftinterfaces, specifically
+      // skip the module adjacent to the interface, but use the caches if
+      // they're present.
+      shouldLoadAdjacentModule = false;
+      break;
+    case ModuleLoadingMode::PreferSerialized:
+      // The rest of the function should be covered by this.
+      break;
+    case ModuleLoadingMode::OnlySerialized:
+      llvm_unreachable("parseable module loader should not have been created");
+    }
+
+    // First, check the cached module path. Whatever's in this cache represents
+    // the most up-to-date knowledge we have about the module.
+    if (fs.exists(cachedOutputPath)) {
+      auto bufOrErr = fs.getBufferForFile(cachedOutputPath);
```

Comment:
Don't check for existence separately; look at the error code if you can't read the file.